checker/match.go

```diff
@@ -0,0 +1,161 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"fmt"
+	"regexp"
+
+	authz "github.com/envoyproxy/data-plane-api/api/auth"
+
+	"github.com/projectcalico/app-policy/proto"
+	"github.com/projectcalico/libcalico-go/lib/selector"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const SPIFFE_ID_PATTERN = "^spiffe://[^/]+/ns/([^/]+)/sa/([^/]+)$"
+
+var spiffeIdRegExp *regexp.Regexp
+
+// match checks if the Rule matches the request.  It returns true if the Rule matches, false otherwise.
+func match(rule *proto.Rule, req *authz.CheckRequest) bool {
+	log.Debugf("Checking rule %v on request %v", rule, req)
+	attr := req.GetAttributes()
+	return matchSource(rule, attr.GetSource()) && matchRequest(rule, attr.GetRequest())
+}
+
+func matchSource(r *proto.Rule, peer *authz.AttributeContext_Peer) bool {
+	// TODO IPSets
+	return matchServiceAccounts(r.GetSrcServiceAccount(), peer)
+}
+
+func matchRequest(rule *proto.Rule, req *authz.AttributeContext_Request) bool {
+	log.WithFields(log.Fields{
+		"request": req,
+	}).Debug("Matching request.")
+	return matchHTTP(rule.GetHttp(), req.GetHttp())
+}
+
+func matchServiceAccounts(saMatch *proto.ServiceAccountSelector, peer *authz.AttributeContext_Peer) bool {
+	principle := peer.GetPrincipal()
+	labels := peer.GetLabels()
+	log.WithFields(log.Fields{
+		"peer":   principle,
+		"labels": labels,
+		"rule":   saMatch},
+	).Debug("Matching service account.")
+	accountName, namespace, err := parseSpiffeId(principle)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"principle": principle,
+			"msg":       err,
+		}).Warn("Unable to parse authenticated principle as SPIFFE ID.")
+		return false
+	}
+	log.WithFields(log.Fields{
+		"name":      accountName,
+		"namespace": namespace,
+	}).Debug("Parsed SPIFFE ID.")
+	if saMatch == nil {
+		log.Debug("nil ServiceAccountMatch.  Return true.")
+		return true
+	}
+	return matchServiceAccountName(saMatch.GetNames(), accountName) &&
+		matchServiceAccountLabels(saMatch.GetLabelSelector(), labels)
+}
+
+// Parse an Istio SPIFFE ID and extract the service account name and namespace.
+func parseSpiffeId(id string) (name, namespace string, err error) {
+	// Init the regexp the first time this is called, and store it in the package namespace.
+	if spiffeIdRegExp == nil {
+		// We drop the returned error here, since we are compiling
+		spiffeIdRegExp, _ = regexp.Compile(SPIFFE_ID_PATTERN)
+	}
+	match := spiffeIdRegExp.FindStringSubmatch(id)
+	if match == nil {
```

Comment:
not sure I see much value in named return in this function. I think this is more idiomatic: 
```go 
match := spiffeIdRegExp.FindStringSubmatch(id)
if match == nil {
    return "", "", fmt.Errorf(...)
    }

return match[2], match[1], nil
}

Reply:
The value is that the function signature is much more clear about which string result is the name and which is the namespace. I originally wrote the function without named return values, then promptly mixed them up the first time I called this function.